Using nested Map in design generates code with compilation issues

## Expected Behavior
Generated code to compile successfully

## Actual Behavior
Generated code fails with a compilation error

## Steps to Reproduce the Problem (including complete and simple design if relevant)
The following simply snippet can be used to reproduce the issue
```golang
package design

import (
	. "goa.design/goa/http/design"
	. "goa.design/goa/http/dsl"
)

var _ = API("dummy", func() {
	Title(`Dummy API`)
})

var SomeRequest = Type("ShowItem", func() {
	Attribute("id", UInt, "id of something.", func() {
		Example(20)
	})
	Attribute("nested_map", MapOf(String, ArrayOf(MapOf(String, Any))), "The nested map")
	Required("id")
})

var _ = Service("dummy", func() {
	Description("Foo")
	HTTP(func() {
		Path("/api/foo")
	})

	Method("Show", func() {
		Description("Show.")
		Payload(SomeRequest)

		HTTP(func() {
			GET("/")
			Response(StatusOK)
		})
	})

})
```

Generate the code and build the generated server type
```
go build gen/http/dummy/server/types.go
```

and this produces the following compilation error
```
gen/http/dummy/server/types.go:39:8: invalid operation: tv[i] (type interface {} does not support indexing)
```

Here is the relevant snippet from the generated code:

```golang
// NewShowShowItem builds a dummy service Show endpoint payload.
func NewShowShowItem(body *ShowRequestBody) *dummysvc.ShowItem {
	v := &dummysvc.ShowItem{
		ID: *body.ID,
	}
	if body.NestedMap != nil {
		v.NestedMap = make(map[string][]map[string]interface{}, len(body.NestedMap))
		for key, val := range body.NestedMap {
			tk := key
			tv := make([]map[string]interface{}, len(val))
			for i, val := range val {
				tv[i] = make(map[string]interface{}, len(val))
				for key, val := range val {
					tk := key
					tv := val
					tv[i][tk] = tv
				}
			}
			v.NestedMap[tk] = tv
		}
	}
	return v
}
```

## Specifications

  - goa Version: v2
  - Platform (Linux/Windows etc.): Mac

--- a/codegen/transform.go
+++ b/codegen/transform.go
@@ -322,12 +322,30 @@ func transformMap(source, target *design.Map, newVar bool, a targs) (string, err
 		"TargetPkg":   a.targetPkg,
 		"Unmarshal":   a.unmarshal,
 		"Scope":       a.scope,
+		"KeyLoopVar":  "",
+		"ValLoopVar":  "",
+	}
+	depth := 0
+	if mapDepth(target.KeyType.Type, &depth); depth > 0 {
+		data["KeyLoopVar"] = string(105 + depth)
+	}
+	depth = 0
+	if mapDepth(target.ElemType.Type, &depth); depth > 0 {
+		data["ValLoopVar"] = string(105 + depth)
 	}
 	var buf bytes.Buffer
 	if err := transformMapT.Execute(&buf, data); err != nil {
 		panic(err) // bug
 	}
 	return buf.String(), nil
+}
+
+func mapDepth(m design.DataType, depth *int) {
+	if mp := design.AsMap(m); mp != nil {
+		*depth++
+		mapDepth(mp.ElemType.Type, depth)
+	}
+	return
 }
 
 func transformAttributeHelpers(source, target design.DataType, a thargs, seen ...map[string]*TransformFunctionData) ([]*TransformFunctionData, error) {
@@ -561,8 +579,8 @@ for {{ .LoopVar }}, val := range {{ .Source }} {
 
 const transformMapTmpl = `{{ .Target }} {{ if .NewVar }}:{{ end }}= make(map[{{ .KeyTypeRef }}]{{ .ElemTypeRef }}, len({{ .Source }}))
 for key, val := range {{ .Source }} {
-	{{ transformAttribute .SourceKey .TargetKey "key" "tk" .SourcePkg .TargetPkg  .Unmarshal true .Scope -}}
-	{{ transformAttribute .SourceElem .TargetElem "val" "tv" .SourcePkg .TargetPkg .Unmarshal true .Scope -}}
-	{{ .Target }}[tk] = tv
+	{{ transformAttribute .SourceKey .TargetKey "key" (printf "tk%s" .KeyLoopVar) .SourcePkg .TargetPkg  .Unmarshal true .Scope -}}
+	{{ transformAttribute .SourceElem .TargetElem "val" (printf "tv%s" .ValLoopVar) .SourcePkg .TargetPkg .Unmarshal true .Scope -}}
+	{{ .Target }}[{{ printf "tk%s" .KeyLoopVar }}] = {{ printf "tv%s" .ValLoopVar }}
 }
 `

--- a/codegen/transform_test.go
+++ b/codegen/transform_test.go
@@ -13,6 +13,8 @@ var (
 	SuperObj      = require(object("a", design.String, "b", design.Int, "c", design.Boolean), "a")
 	SimpleArray   = array(design.String)
 	SimpleMap     = mapa(design.String, design.Int)
+	NestedMap     = mapa(design.String, SimpleMap.Type)
+	NestedMap2    = mapa(design.String, NestedMap.Type)
 	ArrayObj      = object("a", design.String, "b", SimpleArray.Type)
 	ArrayObj2     = object("a", design.String, "b", array(ArrayObj.Type).Type)
 	CompositeObj  = defaulta(require(object("aa", SimpleArray.Type, "bb", SimpleObj.Type), "bb"), "aa", []string{"foo", "bar"})
@@ -54,10 +56,13 @@ func TestGoTypeTransform(t *testing.T) {
 		// simple array and map
 		{"array-unmarshal", SimpleArray, SimpleArray, true, "", arrayCode},
 		{"map-unmarshal", SimpleMap, SimpleMap, true, "", mapCode},
+		{"nested-map-unmarshal", NestedMap, NestedMap, true, "", nestedMapCode},
 		{"object-array-unmarshal", ArrayObj, ArrayObj, true, "", arrayObjUnmarshalCode},
 
 		{"array-marshal", SimpleArray, SimpleArray, false, "", arrayCode},
 		{"map-marshal", SimpleMap, SimpleMap, false, "", mapCode},
+		{"nested-map-marshal", NestedMap, NestedMap, false, "", nestedMapCode},
+		{"nested-map-depth-2-marshal", NestedMap2, NestedMap2, false, "", nestedMap2Code},
 		{"object-array-marshal", ArrayObj, ArrayObj, false, "", arrayObjCode},
 		{"object-array-array-marshal", ArrayObj2, ArrayObj2, false, "", arrayObj2Code},
 
@@ -316,6 +321,41 @@ const mapCode = `func transform() {
 		tk := key
 		tv := val
 		target[tk] = tv
+	}
+}
+`
+
+const nestedMapCode = `func transform() {
+	target := make(map[string]map[string]int, len(source))
+	for key, val := range source {
+		tk := key
+		tvj := make(map[string]int, len(val))
+		for key, val := range val {
+			tk := key
+			tv := val
+			tvj[tk] = tv
+		}
+		target[tk] = tvj
+	}
+}
+`
+
+const nestedMap2Code = `func transform() {
+	target := make(map[string]map[string]map[string]int, len(source))
+	for key, val := range source {
+		tk := key
+		tvk := make(map[string]map[string]int, len(val))
+		for key, val := range val {
+			tk := key
+			tvj := make(map[string]int, len(val))
+			for key, val := range val {
+				tk := key
+				tv := val
+				tvj[tk] = tv
+			}
+			tvk[tk] = tvj
+		}
+		target[tk] = tvk
 	}
 }
 `

